### PR TITLE
issue #3626:Update the bindings-generator to support the namespace’s binding

### PR DESCRIPTION
**Pull request by samuele3hu**

### CODE REVIEW

File: generator.py [hunk 1]
```diff
@@ -932,6 +932,13 @@ def _deep_iterate(self, cursor, depth=0):
         for node in cursor.get_children():
             # print("%s %s - %s" % (">" * depth, node.displayname, node.kind))
             self._deep_iterate(node, depth + 1)
+    def get_script_namespace_name(self, namespace_class_name):
```
Comment (dumganhar) [inline generator.py]: What about `native_namespace_name_to_script_name` or
`scriptname_from_native`?

File: targets/lua/conversions.yaml [hunk 2]
```diff
@@ -16,6 +16,14 @@ conversions:
     "@std::vector<std::basic_string.*>.>.>": "std::vector<std::string>"
     "@std::vector<int.*>.>": "std::vector<int>"
     "@cocos2d::Map<std::basic_string.*,": "cocos2d::Map<std::string,"
+  script_ns_name:
```
Comment (dumganhar) [inline targets/lua/conversions.yaml]: `namespace_map`  or `ns_map`?

Comment (dumganhar) [inline targets/lua/conversions.yaml]: I like `ns_map`, it's shorter and clearer.

Comment (samuele3hu) [inline targets/lua/conversions.yaml]: Does it need  prefix "script_"

Comment (dumganhar) [inline targets/lua/conversions.yaml]: I don't think so. `ns_map` certainly refers to `native_ns` and `script_ns`.

File: generator.py [hunk 3]
```diff
@@ -932,6 +932,13 @@ def _deep_iterate(self, cursor, depth=0):
         for node in cursor.get_children():
             # print("%s %s - %s" % (">" * depth, node.displayname, node.kind))
             self._deep_iterate(node, depth + 1)
+    def scriptname_from_native(self, namespace_class_name):
+        script_ns_dict = self.config['conversions']['ns_map']
+
+        for (k, v) in script_ns_dict.items():
+            if namespace_class_name.find(k) >= 0:
+                return namespace_class_name.replace("*","").replace("const ", "").replace(k,v)
+        return None
```
Comment (dumganhar) [inline generator.py]: Why to return `None` if can't find the name, if user's module doesn't contain a namespace which in the `ns_map`, the bindings will not work, right?

Comment (dumganhar) [inline generator.py]: I think it should return `namespace_class_name.replace("*","").replace("const ", "")`, right?

``` python
def scriptname_from_native(self, namespace_class_name):
 +        script_ns_dict = self.config['conversions']['ns_map']
 +        ret = namespace_class_name.replace("*","").replace("const ", "");
 +        for (k, v) in script_ns_dict.items():
 +            if namespace_class_name.find(k) >= 0:
 +                return ret.replace(k,v)
 +        return ret
```

Comment (samuele3hu) [inline generator.py]: yes,users should configure the key-value pair in the ns_map, or let it return the name_class_name.

File: generator.py [hunk 4]
```diff
@@ -932,7 +932,20 @@ def _deep_iterate(self, cursor, depth=0):
...
+        for (k, v) in script_ns_dict.items():
+            if namespace_class_name.find(k) >= 0:
+                return namespace_class_name.replace("*","").replace("const ", "").replace(k,v)
+        if namespace_class_name.find("::") >= 0:
+            if namespace_class_name.find("std::") == 0:
+                return namespace_class_name
+            else:
+                raise Exception("%s includes namespace's name "
+                      "but the ns_map field of the conversions field in the conversions.yaml "
+                      "has not corresponding modular name." % namespace_class_name)
```
Comment (dumganhar) [inline generator.py]: Error: The namespace (%s) conversion wasn't set in `ns_map` section of conversions.yaml.

**Comment by dumganhar** [comment]
Great. Thanks.

**Pull request merged**